src/main/java/io/jenkins/plugins/artifact_manager_jclouds/swift/SwiftBlobstore.java

```diff
@@ -0,0 +1,182 @@
+
+import hudson.Extension;
+import hudson.Util;
+import hudson.util.Secret;
+import io.jenkins.plugins.artifact_manager_jclouds.BlobStoreProvider;
+import io.jenkins.plugins.artifact_manager_jclouds.BlobStoreProviderDescriptor;
+
+public class SwiftBlobstore extends BlobStoreProvider {
+
+	private static final long serialVersionUID = 5517446936331565739L;
+
+	private static final Logger LOGGER = Logger.getLogger(SwiftBlobstore.class.getName());
+
+	private transient RegionScopedBlobStoreContext blobstoreContext;
+
+	@DataBoundConstructor
+	public SwiftBlobstore() {
+	}
+
+	@Override
+	public String getPrefix() {
+		return getConfiguration().getPrefix();
+	}
+
+	@Override
+	public String getContainer() {
+		return getConfiguration().getContainer();
+	}
+
+	@Override
+	public boolean isDeleteArtifacts() {
+		return getConfiguration().isDeleteArtifacts();
+	}
+
+	@Override
+	public boolean isDeleteStashes() {
+		return getConfiguration().isDeleteStashes();
+	}
+
+	private SwiftBlobstoreConfig getConfiguration() {
+		return SwiftBlobstoreConfig.get();
+	}
+
+	private String getRegion() {
+		return getConfiguration().getRegion();
+	}
+
+	private RegionScopedBlobStoreContext getRegionContext() {
+		if (blobstoreContext == null) {
+			LOGGER.log(Level.FINER, "Building Swift context");
+
+			SwiftBlobstoreConfig config = getConfiguration();
+			final Properties overrides = new Properties();
+			UsernamePasswordCredentials credential = getConfiguration().getCredential();
+
+			overrides.put(KeystoneProperties.KEYSTONE_VERSION, "3");
+			overrides.put(KeystoneProperties.SCOPE, "project:" + config.getProjectName());
+			if (Util.fixEmptyAndTrim(config.getProjectDomain()) != null) {
+				overrides.put(KeystoneProperties.PROJECT_DOMAIN_NAME, config.getProjectDomain());
+			}
+
+
+			blobstoreContext = ContextBuilder.newBuilder(new SwiftApiMetadata())
+					.credentials(String.format("%s:%s", config.getUserDomain(), credential.getUsername()),
+							Secret.toString(credential.getPassword()))
+					.overrides(overrides).endpoint(config.getEndpoint()).build(RegionScopedBlobStoreContext.class);
+
+			SwiftApi api = blobstoreContext.unwrapApi(SwiftApi.class);
+			api.getContainerApi(getRegion()).create(getContainer());
+			AccountApi accountApi = api.getAccountApi(getRegion());
+			if (!accountApi.get().getTemporaryUrlKey().isPresent()) {
+				accountApi.updateTemporaryUrlKey(UUID.randomUUID().toString());
+			}
+
+		}
+
+		return blobstoreContext;
+	}
+
+	@Override
+	public BlobStoreContext getContext() throws IOException {
+		return getRegionContext();
+	}
+
+	@Override
+	public BlobStore getBlobStore() throws IOException {
+		return getRegionContext().getBlobStore(getRegion());
+	}
+
+	@Override
+	public URI toURI(String container, String key) {
+		try {
+			Blob blob = getBlobStore().blobBuilder(key).build();
+			blob.getMetadata().setContainer(container);
+			return toExternalURL(blob, HttpMethod.GET).toURI();
```

Comment:
This should not be implemented using a presigned URL. It should be a predictable URI without authentication information.

Reply:
Not sure if this is possible.
You can access the blobstore directly only when you have the corresponding permissions configured I think, so no anonymous access.

Reply:
The URI need not be a resolvable URL. It is not in the case of the S3 (incl. MinIO) implementation. [Javadoc](https://javadoc.jenkins.io/jenkins/util/VirtualFile.html#toURI--)

Reply:
(probably the method in this plugin’s API should `@link` to that for clarity)